scripts/libraries/ml/ml/preprocessing.py

```diff
@@ -42,12 +44,32 @@ def __call__(self, *args):
             raise ValueError("Expected batches to be 1")
         if c != 1 and c != 3:
             raise ValueError("Expected channels to be 1 or 3")
+
+        # Place the image buffer at the end of the input buffer so we can convert it in-place.
         pixfmt = image.GRAYSCALE if c == 1 else image.RGB565
-        img = image.Image(w, h, pixfmt, buffer=buffer)
+        offset = len(buffer) - (w * h * (1 if c == 1 else 2))
+        img = image.Image(w, h, pixfmt, buffer=memoryview(buffer)[offset:])
 
         # Copy and scale (if needed) the input image to the input buffer.
         hints = image.BILINEAR | image.CENTER | image.SCALE_ASPECT_EXPAND | image.BLACK_BACKGROUND
         img.draw_image(self._image, 0, 0, roi=self.roi, hint=hints)
 
-        # Scale and convert the image to input tensor data.
-        img.unpack(buffer, dtype, scale=self.scale, mean=self.mean, stdev=self.stdev)
+        # Convert the image in-place into an ndarray input tensor.
+        array = img.ndarray(dtype, buffer=buffer)
+
+        # Normalize the input tensor.
+        if dtype == ord('f'):
+            fscale = (self.scale[1] - self.scale[0]) / 255.0
+            fadd = self.scale[0]
+
+            def grayscale(x):
+                return (x[0] * 0.299) + (x[1] * 0.587) + (x[2] * 0.114)
+
+            if c == 1:
+                fadd = (fadd - grayscale(self.mean)) / grayscale(self.stdev)
+                fscale = fscale / grayscale(self.stdev)
+            else:
+                fadd = (fadd - np.array(self.mean)) / np.array(self.stdev)
+                fscale = fscale / np.array(self.stdev)
+
+            array = (array * fscale) + fadd
```

Comment:
By switching to the ndarray here all the complex math was moved outside into the python code. Much nicer.